gitpod-io/gitpod
# [supervisor] execve into ring3

This PR makes ring2 execve `supervisor run` (aka _ring3_). This way PID 1 in the workspace is `supervisor run`, much like it would be without user namespaces. Hence supervisor can properly do its job as reaper, and the graceful shutdown of tasks is a bit less brittle (#2654).

### How does it work?
Instead of starting a child process, ring2 now calls setgid/setuid to drop to the Gitpod user (`33333`) and then execve's `/proc/self/exe run`, thereby replacing the ring2 process with supervisor run.

### Why didn't we do this before?
setuid/setgid have a [long and difficult history in Go](https://github.com/golang/go/issues/1435). Go 1.16 introduced `AllThreadsSyscall` which makes setuid/setgid work on all threads without the use of CGO. TLDR; Before Go 1.16 we'd need CGO, but we don't want CGO because supervisor has to run in glibc and muslibc environments. [cap](https://pkg.go.dev/kernel.org/pub/linux/libs/security/libcap/cap) - a Go-native port of libcap - supports this feature, and if it's present supports capability-preserving `setuid/setgid` calls without CGO.

### But wait, Go 1.16 is still in beta?!?
Indeed. That's why leeway now supports choosing a Go version on a package level. Starting with this PR, we compile supervisor with [go1.16beta1](https://golang.org/dl/#go1.16beta1), and the rest using the Go version we have installed in the dev-environment image. Once [Go 1.16](https://tip.golang.org/doc/go1.16) is out of beta (Feb 2021), we'll switch over for all components. Go betas have been really rather stable and I don't expect issues because we build supervisor using Go 1.16. Note https://github.com/golang/go/issues/43149 does not concern us, because only ring2 uses AllThreadsSyscall and ring2 does not use `signal`.

### How to test
1. Enable feature preview
2. Start a workspace on http://cw-supervisor-execve-ring3.staging.gitpod-dev.com/#https://github.com/csweichel/test-repo/tree/proper-shutdown. This workspace starts a command which listens for sigterm and writes a file.
3. Stop the workspace. You should see `proper-shutdown.txt` in the list of changed files. This file was created when the task stopped properly.
4. Create a zombie: `./zombie & sleep 1 && ps guaxf | grep zombie` ... once the zombie parent exits `ps guaxf | grep defunct` should not return anything. Previously, the defunct process would still have been around.

https://user-images.githubusercontent.com/3210701/103635228-80148480-4f48-11eb-8672-a282aecc9e9b.mov

/werft https
